dask/base.py

```diff
@@ -264,17 +265,19 @@ def visualize(*args, **kwargs):
 
 
 function_cache = {}
+function_cache_lock = threading.Lock()
 
 
 def normalize_function(func):
     try:
         return function_cache[func]
     except KeyError:
         result = _normalize_function(func)
-        if len(function_cache) >= 500:  # clear half of cache if full
-            for k in list(function_cache)[::2]:
-                del function_cache[k]
-        function_cache[func] = result
+        with function_cache_lock:
+            if len(function_cache) >= 500:  # clear half of cache if full
```

Comment:
Actually, nvm, this adds a race condition.

1. Thread1 sees cache is full
2. Thread2 gets control, sees cache is full, gets lock and clears it
3. Thread1 regains control, goes to clear cache, clears it again (shouldn't error, but is unintended)

To fix this but still avoid the lock you might add a repeat `len` check inside the lock as well (with a comment):

```python
if len(function_cache) >= 500:
    with function_cache_lock:
        # Check again to avoid race condition
        if len(function_cache) >= 500:
            for k in list(function_cache)[::2]:
                del function_cache[k]
```